### DISCUSSION TITLE
How to access the Pareto front in the `full_objective_function`

### DISCUSSION THREAD

**Discussion by randomx207** [MilesCranmer/PySR#523] [Q&A]
Hi,

Thank you for developing this library! I am wondering if there is a simple way to access the Pareto front in the `full_objective_function`. Specifically, I would like to get a list of tuples containing the complexity and loss value for each expression in julia objective function, like [(complexity1, loss1), (complexity2, loss2),...]. This would allow me to evaluate a given expression tree based on the Pareto optimal values.

Additionally, is there a way to pass an expression tree provided by the PyCall into Julia with this library? 

Thanks again!

**Comment by MilesCranmer** [comment]
Hey @randomx207,

Interesting idea. I will note that one assumption in SymbolicRegression.jl is that the loss function is static and does not change over time. For example, I'm not sure how to actually create the Pareto frontier if the loss of any member of the frontier is co-dependent on the rest of the frontier (if that makes sense?). In addition, losses are cached, and if the loss were to change as a function of the global search, this would mean you would need to do many more calculations than are currently used, as you would need to re-compute the loss for any equation when comparing it to other candidates.

However, one option that you could consider is simply re-running the search from a checkpoint (`warm_start=True` and calling `.fit(X, y)` again in PySR, or simply calling `fit!(mach)` on a trained machine in Julia), but with a different loss function each time. Basically: 

1. Run the search with an initial loss function.
2. Create a new loss function based on the Pareto front returned from (1).
3. Change the options to use that loss function, and continue the search.
4. Repeat.

I think this would actually work for you, because restarting a search would automatically force the re-calculation of all cached losses and the Pareto frontier. This would also be a "cleaner" approach compared with depending on the instantaneous Pareto frontier in the loss.

You could tweak `niterations` and `ncycles_per_iteration` based on how often you want to update the loss function with the new Pareto front.

Does that help at all?

Cheers,
Miles

**Reply by randomx207** [reply]
Thank you for the detailed reply! However, I don't think re-calling fit can solve my problem. Let me explain more clearly: What I actually want to do is, every time the `objective_function` is run, Julia will use PyCall to pass an expression to Python. The Python program will then feed this expression into a simulation system that is computationally slower than PySR (and cannot be implemented in Julia). Python will obtain a score `s` and return it to Julia, to serve as an additional reference for the Loss `Loss = f(mse, complexity, s)`. 

The issue is that the Python calls are very slow. Therefore, I want to skip calling the additional Python program for some expressions that clearly have no hope, in order to speed things up. Specifically, first use a simple function `Loss_bound = g(mse, complexity)` to check if there is any hope of surpassing the Pareto frontier of expressions with similar complexity. If not, then do not send it to Python for scoring, but simply assign it a benchmark score `s=s_default`, thereby accelerating the computation.

**Reply by MilesCranmer** [reply]
Thanks for explaining this. It's indeed tricky. It is useful for the genetic algorithm to have accurate losses of even the mediocre expressions so that it has a sense of "direction" in terms of what expressions it should breed more frequently in a given population. So even if you are able to avoid expensive simulation calls on some of them, I worry that it might hurt the overall search efficiency as it would only have accurate losses of the best expressions (which it might need to randomly stumble upon, rather than selectively breed into existence).

However if you still want to proceed, I do think that re-calling fit with an updated loss function would work though. Sure it would not be completely up-to-date for the pareto front, but it would get you 80% of the way there for 20% of the effort :)


Another option is to do what we did in https://arxiv.org/abs/2006.11287v2 and basically fit a neural network (or other flexible model) to your simulation system first. And then fit PySR to the neural network, as you can simply generate input->output points from the neural net and fit using MSE (which is quite fast). See https://github.com/MilesCranmer/PySR/blob/master/examples/pysr_demo.ipynb for an example of this at the end.

**Reply by randomx207** [reply]
Thanks for the suggestion! I'll give that a try and see if it helps resolve the issue I'm running into.

**Comment by randomx207** [comment]
Hi @MilesCranmer, 

I would also like to ask: what should I do if I want to generate an initial population (i.e., a sympy expression list, or preorder traversal token list, whatever) in Python, and pass it to julia during the pysr population initialization phase? I have looked at https://github.com/MilesCranmer/PySR/discussions/491 and have clone the source code, but I want to know:
1. Is it possible to modify the function `function Population` to pass expressions from PyCall?
2. If so, how do I build a pysr tree from python's expression objects? I don't know julia well. Can you help me?

for example, what I want to do is like this

```
    return Population{T,L}(
        [
            PopMember(
                dataset,
                call_python_and_get_a_tree(),       # here
                options;
                parent=-1,
                deterministic=options.deterministic,
            ) for i in 1:population_size
        ],
```
